File: NCPC/2014/ncpc2014/basincity/submissions/accepted/mdr-simplicial.cpp
// Solution by Markus Dregi
//
// If it finds a simplicial vertex, it adds it to the solution. Otherwise it
// finds a lowest degree neighbour and branches on it.
//
// Complexity: <= 5^k n


#include<iostream>
#include<queue>
#include<vector>
#include<string>
#include<cmath>

#define IN 0
#define OUT 1
#define FREE 2

using namespace std;

vector< vector<int> > graph;
vector< vector<bool> > adjacent;
vector<int> degree;
vector<int> status;

int k, n;

// Finds a good vertex to branch on. Either a simplicial vertex or a low-degree
void search(int& bv, bool& simplicial)
{
    simplicial = false;
    bv = -10;
    int min_deg = 10;

    for(int i = 0; i < n; ++i) {
        if(status[i] != FREE)
            continue;

        bool s = true;
        int d = 0;
        for(int j = 0; j < graph[i].size(); ++j) {
            if(status[graph[i][j]] != FREE)
                continue;
            ++d;

            for(int k = j+1; k < graph[i].size() && s; ++k)
                s &= (status[graph[i][k]] != FREE ||
                        adjacent[graph[i][j]][graph[i][k]]);

        }

        if(s) {
            bv = i;
            simplicial = true;
            return;
        } else if(d < min_deg) {
            bv = i;
            min_deg = d;
        }
    }
}

int branch(const vector<int>& branch_vertices, const int& target);
int solve(const int& target);

// Adds branch_vertices to the solution and searches for a solution of size
// target.
int branch(const vector<int>& branch_vertices, const int& target)
{
    // Locate the free neighbours of the branching vertex
    vector<int> new_outs;

    for(int i = 0; i < branch_vertices.size(); ++i)
        for(int j = 0; j < graph[branch_vertices[i]].size(); ++j)
            if(status[graph[branch_vertices[i]][j]] == FREE)
                new_outs.push_back(graph[branch_vertices[i]][j]);

    // Update status
    for(int i = 0; i < new_outs.size(); ++i)
        status[new_outs[i]] = OUT;
    for(int i = 0; i  < branch_vertices.size(); ++i)
        status[branch_vertices[i]] = IN;

    // Recurse
    int size = branch_vertices.size();
    size += solve(target-size);

    // Reset status
    for(int i = 0; i < new_outs.size(); ++i)
        status[new_outs[i]] = FREE;
    for(int i = 0; i  < branch_vertices.size(); ++i)
        status[branch_vertices[i]] = FREE;

    return size;
}

// Searches for a solution of size target
int solve(const int& target)
{
    if(target <= 0)
        return 0;

    bool simplicial;
    int bv;
    search(bv, simplicial);

    if(bv < 0)
        return 0;

    vector<int> branch_set(1, bv);
    int best = branch(branch_set, target);

    if(!simplicial) {
        for(int i = 0; i < graph[bv].size() && best < target; ++i) {
            if(status[graph[bv][i]] == FREE) {
                branch_set[0] = graph[bv][i];
                best = max(best, branch(branch_set, target));
            }
        }
    }

    return best;
}

int main()
{
    // Read Input
    cin >> k >> n;

    if(n >= 5*k-4) {
        cout << "possible" << endl;
        return 0;
    }

    graph = vector< vector<int> >(n);
    adjacent = vector< vector<bool> >(n, vector<bool>(n, false));
    degree = vector<int>(n);
    status = vector<int>(n, FREE);

    for(int i = 0; i < n; ++i) {
        cin >> degree[i];

        for(int j = 0; j < degree[i]; ++j) {
            int nghbr;
            cin >> nghbr;
            graph[i].push_back(nghbr-1);
            adjacent[i][nghbr-1] = true;
        }
    }

    // Output
    cout << (solve(k) >= k ? "possible" : "impossible") << endl;

    return 0;
}

File: NCPC/2014/ncpc2014/basincity/submissions/accepted/mdr-simplicial.py
# Solution by Markus Dregi
# Same algorithm as mdr-simplicial-fast.cpp
# Complexity: 3^k n

import sys
from sys import stdin

# Finds a good vertex to branch on. Either a simplicial vertex or a low-degree
def search():
    bv = -10
    min_deg = 10
    for v in range(n):
        if used[v]:
            continue
        s = True
        d = 0
        for i in range(len(graph[v])):
            n1 = graph[v][i]
            if used[n1]:
                continue
            d += 1
            for n2 in graph[v][i+1:]:
                if not s:
                    break
                if not used[n2] and not adjacent[n1][n2]:
                    s = False
        if s:
            return True, v
        elif d < min_deg:
            bv = v
            min_deg = d
    return False, bv

# Adds all vertices in branch_vertices to the independent set and 
# recurses for a solution of size target.
def branch(branch_vertices, target):
    new_used = []
    for u in branch_vertices:
        for v in graph[u]:
            if not used[v]:
                new_used.append(v)

    # Update
    for v in new_used+branch_vertices:
        used[v] = True

    # Recurse
    size = len(branch_vertices)
    size += solve(target-size)

    # Reset
    for v in new_used+ branch_vertices:
        used[v] = False

    return size

# Searches for a solution of size target
def solve(target):
    if target <= 0:
        return 0

    simplicial, bv = search()
    if bv < 0:
        return 0

    best = branch([bv], target)
    if not simplicial:
        for i in range(len(graph[bv])):
            u = graph[bv][i]
            if used[u]:
                continue
            for v in graph[bv][i+1:]:
                if not used[v] and not adjacent[u][v]:
                    best = max(best, branch([u,v], target))
                    if best >= target:
                        return best

    return best

d = int(stdin.readline())
n = int(stdin.readline())

if d is 0 or n >= 5*d-4:
    print "possible"
    sys.exit()

graph = [[] for i in range(n)]
adjacent = [[False for i in range(n)] for j in range(n)]
used = [False]*n

for i in range(n):
    s = stdin.readline().split()
    for nbr in s[1:]:
        graph[i].append(int(nbr)-1)
        adjacent[i][int(nbr)-1] = True

print ("possible" if solve(d) >= d else "impossible")
